internal/server/evaluation/evaluator.go

```diff
@@ -0,0 +1,440 @@
+}
+
+// Evaluator is an implementation of the MultiVariateEvaluator.
+type Evaluator struct {
+	logger *zap.Logger
+	store  EvaluationStorer
+}
+
+// NewEvaluator is the constructor for an Evaluator.
+func NewEvaluator(logger *zap.Logger, store EvaluationStorer) *Evaluator {
+	return &Evaluator{
+		logger: logger,
+		store:  store,
+	}
+}
+
+// MultiVariateEvaluator is an abstraction for evaluating a flag against a set of rules for multi-variate flags.
+type MultiVariateEvaluator interface {
+	Evaluate(ctx context.Context, r *flipt.EvaluationRequest) (*flipt.EvaluationResponse, error)
+}
+
+func (e *Evaluator) Evaluate(ctx context.Context, r *flipt.EvaluationRequest) (resp *flipt.EvaluationResponse, err error) {
+	var (
+		startTime     = time.Now().UTC()
+		namespaceAttr = metrics.AttributeNamespace.String(r.NamespaceKey)
+		flagAttr      = metrics.AttributeFlag.String(r.FlagKey)
+	)
+
+	metrics.EvaluationsTotal.Add(ctx, 1, metric.WithAttributeSet(attribute.NewSet(namespaceAttr, flagAttr)))
+
+	defer func() {
+		if err == nil {
+			metrics.EvaluationResultsTotal.Add(ctx, 1,
+				metric.WithAttributeSet(
+					attribute.NewSet(
+						namespaceAttr,
+						flagAttr,
+						metrics.AttributeMatch.Bool(resp.Match),
+						metrics.AttributeSegment.String(resp.SegmentKey),
+						metrics.AttributeReason.String(resp.Reason.String()),
+						metrics.AttributeValue.String(resp.Value),
+					),
+				),
+			)
+		} else {
+			metrics.EvaluationErrorsTotal.Add(ctx, 1, metric.WithAttributeSet(attribute.NewSet(namespaceAttr, flagAttr)))
+		}
+
+		metrics.EvaluationLatency.Record(
+			ctx,
+			float64(time.Since(startTime).Nanoseconds())/1e6,
+			metric.WithAttributeSet(
+				attribute.NewSet(
+					namespaceAttr,
+					flagAttr,
+				),
+			),
+		)
+	}()
+
+	resp = &flipt.EvaluationResponse{
+		RequestId:      r.RequestId,
+		EntityId:       r.EntityId,
+		RequestContext: r.Context,
+		FlagKey:        r.FlagKey,
+		NamespaceKey:   r.NamespaceKey,
+	}
+
+	flag, err := e.store.GetFlag(ctx, r.NamespaceKey, r.FlagKey)
+	if err != nil {
+		resp.Reason = flipt.EvaluationReason_ERROR_EVALUATION_REASON
+
+		var errnf errs.ErrNotFound
+		if errors.As(err, &errnf) {
+			resp.Reason = flipt.EvaluationReason_FLAG_NOT_FOUND_EVALUATION_REASON
+		}
+
+		return resp, err
+	}
+
+	if !flag.Enabled {
+		resp.Match = false
+		resp.Reason = flipt.EvaluationReason_FLAG_DISABLED_EVALUATION_REASON
+		return resp, nil
+	}
+
+	rules, err := e.store.GetEvaluationRules(ctx, r.NamespaceKey, r.FlagKey)
+	if err != nil {
+		resp.Reason = flipt.EvaluationReason_ERROR_EVALUATION_REASON
+		return resp, err
+	}
+
+	if len(rules) == 0 {
+		e.logger.Debug("no rules match")
+		return resp, nil
+	}
+
+	var lastRank int32
+
+	// rule loop
+	for _, rule := range rules {
+		if rule.Rank < lastRank {
+			resp.Reason = flipt.EvaluationReason_ERROR_EVALUATION_REASON
+			return resp, errs.ErrInvalidf("rule rank: %d detected out of order", rule.Rank)
+		}
```

Comment:
Thinking out loud (I would keep this as is for now).

After our discussion the other day on how two folks updating rules at the same time (one create and one re-order) would lead to two rules getting the same rank. This would cause the flag to always fail evaluation here.

Multiple thoughts come to mind at once:
1. We should open an issue for the ordering bug
2. Should this error be internal. Seems that while rank is controlled by users, the field is actually managed by the Flipt server (order takes IDs and set appropriate rank). So failure here is because Flipt messed up, not because the user put in something invalid.

cc @markphelps

Reply:
@GeorgeMac agreed on both counts